File: sdk/java/src/main/java/com/hjtpx/captcha/retry/RetryManager.java
package com.hjtpx.captcha.retry;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.Callable;

public class RetryManager {
    private static final Logger logger = LoggerFactory.getLogger(RetryManager.class);
    private final RetryConfig config;

    public RetryManager() {
        this(new RetryConfig());
    }

    public RetryManager(RetryConfig config) {
        this.config = config;
    }

    public <T> T execute(Callable<T> callable) throws Exception {
        int attempt = 0;
        Exception lastException = null;

        while (attempt <= config.getMaxRetries()) {
            try {
                return callable.call();
            } catch (Exception e) {
                lastException = e;
                attempt++;

                if (attempt > config.getMaxRetries()) {
                    logger.warn("Max retries ({}) reached, giving up", config.getMaxRetries());
                    break;
                }

                if (!shouldRetry(e)) {
                    logger.warn("Exception not retryable, giving up: {}", e.getMessage());
                    break;
                }

                long delay = config.calculateDelay(attempt - 1);
                logger.info("Retry attempt {}/{}, waiting {}ms", attempt, config.getMaxRetries(), delay);

                try {
                    Thread.sleep(delay);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw e;
                }
            }
        }

        throw lastException;
    }

    private boolean shouldRetry(Exception e) {
        return config.isRetryableException(e) || isNetworkException(e);
    }

    private boolean isNetworkException(Exception e) {
        String message = e.getMessage();
        if (message == null) {
            return false;
        }
        String lowerMessage = message.toLowerCase();
        return lowerMessage.contains("timeout") ||
               lowerMessage.contains("connection reset") ||
               lowerMessage.contains("connection refused") ||
               lowerMessage.contains("socket hang up") ||
               lowerMessage.contains("network");
    }

    public RetryConfig getConfig() {
        return config;
    }
}

File: sdk/csharp/src/Hjtpx.Captcha.Sdk/Retry/RetryManager.cs
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Hjtpx.Captcha.Retry;

public class RetryManager
{
    private readonly RetryConfig _config;
    private readonly ILogger<RetryManager> _logger;

    public RetryManager(RetryConfig? config = null, ILogger<RetryManager>? logger = null)
    {
        _config = config ?? new RetryConfig();
        _logger = logger ?? NullLogger<RetryManager>.Instance;
    }

    public async Task<T> ExecuteAsync<T>(Func<Task<T>> func, CancellationToken cancellationToken = default)
    {
        int attempt = 0;
        Exception? lastException = null;

        while (attempt <= _config.MaxRetries)
        {
            try
            {
                return await func();
            }
            catch (Exception e)
            {
                lastException = e;
                attempt++;

                if (attempt > _config.MaxRetries)
                {
                    _logger.LogWarning("Max retries ({MaxRetries}) reached, giving up", _config.MaxRetries);
                    break;
                }

                if (!ShouldRetry(e))
                {
                    _logger.LogWarning("Exception not retryable, giving up: {Message}", e.Message);
                    break;
                }

                long delay = _config.CalculateDelay(attempt - 1);
                _logger.LogInformation("Retry attempt {Attempt}/{MaxRetries}, waiting {Delay}ms", attempt, _config.MaxRetries, delay);

                await Task.Delay((int)delay, cancellationToken);
            }
        }

        throw lastException!;
    }

    public T Execute<T>(Func<T> func)
    {
        int attempt = 0;
        Exception? lastException = null;

        while (attempt <= _config.MaxRetries)
        {
            try
            {
                return func();
            }
            catch (Exception e)
            {
                lastException = e;
                attempt++;

                if (attempt > _config.MaxRetries)
                {
                    _logger.LogWarning("Max retries ({MaxRetries}) reached, giving up", _config.MaxRetries);
                    break;
                }

                if (!ShouldRetry(e))
                {
                    _logger.LogWarning("Exception not retryable, giving up: {Message}", e.Message);
                    break;
                }

                long delay = _config.CalculateDelay(attempt - 1);
                _logger.LogInformation("Retry attempt {Attempt}/{MaxRetries}, waiting {Delay}ms", attempt, _config.MaxRetries, delay);

                Thread.Sleep((int)delay);
            }
        }

        throw lastException!;
    }

    private bool ShouldRetry(Exception e)
    {
        return _config.IsRetryableException(e) || IsNetworkException(e);
    }

    private bool IsNetworkException(Exception e)
    {
        string? message = e.Message?.ToLowerInvariant();
        if (string.IsNullOrEmpty(message))
        {
            return false;
        }

        return message.Contains("timeout") ||
               message.Contains("connection reset") ||
               message.Contains("connection refused") ||
               message.Contains("socket hang up") ||
               message.Contains("network");
    }
}
